mlflow/pyfunc/__init__.py

```diff
@@ -724,6 +728,80 @@ def load_model(
     return PyFuncModel(model_meta=model_meta, model_impl=model_impl)
 
 
+def _download_model_conda_env(model_uri):
+    conda_yml_file_name = _get_flavor_configuration_from_uri(model_uri, FLAVOR_NAME)[ENV]
+    return _download_artifact_from_uri(append_to_uri_path(model_uri, conda_yml_file_name))
+
+
+def get_model_dependencies(model_uri, format="pip"):  # pylint: disable=redefined-builtin
+    """
+    Given a model URI, and format ("pip" or "conda", default value "pip"),
+    return the downloaded dependency file path.
+    If "pip" format specified but model does not have "requirements.txt" file,
+    fallback to parse the pip section of the model's "conda.yaml" and ignore other
+    non-pip dependencies in the "conda.yaml" file.
+    """
+    req_file_uri = append_to_uri_path(model_uri, _REQUIREMENTS_FILE_NAME)
+
+    if format == "pip":
+        try:
+            pip_file_path = _download_artifact_from_uri(req_file_uri)
+        except Exception as e:
+            # fallback to download conda.yaml file and parse the "pip" section from it.
+            _logger.info(
+                f"Download model '{_REQUIREMENTS_FILE_NAME}' file failed, error is {repr(e)}. "
+                "Fallback to fetch pip requirements from the model's 'conda.yaml' file and "
+                "other conda dependencies will be ignored."
+            )
+            conda_yml_path = _download_model_conda_env(model_uri)
+            pip_deps = None
+            conda_deps = []
+            try:
+                with open(conda_yml_path, "r") as yf:
+                    conda_yml = yaml.safe_load(yf)
+
+                for dep in conda_yml["dependencies"]:
+                    if isinstance(dep, str):
+                        conda_deps.append(dep)
+                    if isinstance(dep, dict) and "pip" in dep:
+                        pip_deps = dep["pip"]
+            except Exception as e:
+                raise MlflowException(
+                    f"Parse conda.yaml file in the model directory failed, error: {repr(e)}."
+                )
```

Comment:
Do we need this try-catch?

Reply:
I preferred add this try catch, so we can wrap the original exception with more clear error message.

Reply:
This is how the error message looks like:

```python
2022/03/29 11:25:20 INFO mlflow.pyfunc: Download model 'requirements.txt' file failed, error is OSError("No such file or directory: '/var/folders/33/kkm875tn7cxb89b8mmcykrx00000gn/T/tmpu3scvbl2/requirements.txt'"). Fallback to fetch pip requirements from the model's 'conda.yaml' file and other conda dependencies will be ignored.
Traceback (most recent call last):
  File "/Users/harutakakawamura/Desktop/repositories/mlflow/mlflow/pyfunc/__init__.py", line 749, in get_model_dependencies
    pip_file_path = _download_artifact_from_uri(req_file_uri)
  File "/Users/harutakakawamura/Desktop/repositories/mlflow/mlflow/tracking/artifact_utils.py", line 96, in _download_artifact_from_uri
    artifact_path=artifact_path, dst_path=output_path
  File "/Users/harutakakawamura/Desktop/repositories/mlflow/mlflow/store/artifact/local_artifact_repo.py", line 79, in download_artifacts
    raise IOError("No such file or directory: '{}'".format(local_artifact_path))
OSError: No such file or directory: '/var/folders/33/kkm875tn7cxb89b8mmcykrx00000gn/T/tmpu3scvbl2/requirements.txt'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/harutakakawamura/Desktop/repositories/mlflow/mlflow/pyfunc/__init__.py", line 764, in get_model_dependencies
    for dep in conda_yml["dependencies"]:
KeyError: 'dependencies'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "a.py", line 39, in <module>
    test_get_model_dependencies_read_conda_file(Path(tmp_path))
  File "a.py", line 33, in test_get_model_dependencies_read_conda_file
    Path(mlflow.pyfunc.get_model_dependencies(model_path, format="pip")).read_text()
  File "/Users/harutakakawamura/Desktop/repositories/mlflow/mlflow/pyfunc/__init__.py", line 771, in get_model_dependencies
    f"Parse conda.yaml file in the model directory failed, error: {repr(e)}."
mlflow.exceptions.MlflowException: Parse conda.yaml file in the model directory failed, error: KeyError('dependencies').
```